### PR TITLE
Nuno Amado

### PR DESCRIPTION
**Pull request by jlouros**

### CODE REVIEW

File: Changes_NunoAmado.txt [hunk 1]
```diff
@@ -0,0 +1,32 @@
+All the requirements complete
```
Comment (jlouros) [inline Changes_NunoAmado.txt]: Good organisation 👌

File: src/Microservice.Api/Contracts/FundDetails.cs [hunk 2]
```diff
@@ -110,4 +122,73 @@ public class Yield
         public string amount { get; set; }
     }
 
+
+    public class TotalSumAttribute : ValidationAttribute
```
Comment (jlouros) [inline src/Microservice.Api/Contracts/FundDetails.cs]: Like this approach 👌

File: src/Microservice.Api/Controllers/FundsController.cs [hunk 3]
```diff
@@ -68,25 +78,174 @@ public IActionResult Post([FromBody] object payload)
...
+                        f.Provider = _context.Providers.First(x => x.Code == f.Provider.Code);
 
-                return Created("http://localhost:5000/api/funds", fundsDetails.id);
+                    _context.Funds.Add(f);
+                    var t = _context.SaveChangesAsync().Result;
+                    //_context.SaveChanges();
+
+                    #endregion
+
+                    return Created("https://localhost:44329/api/funds", fundsDetails.id);
```
Comment (jlouros) [inline src/Microservice.Api/Controllers/FundsController.cs]: hardcoded Uri

File: src/Microservice.Api/Controllers/ImportController.cs [hunk 4]
```diff
@@ -29,17 +29,40 @@ public async Task<IActionResult> Post([FromBody] JArray payload)
...
+                    var m = fundMapper.FromContractToDomain(fd);
+                    _context.Funds.Add(m);
+
+                }
+
                 _context.SaveChanges();
             }
+            catch (Exception ex)
+            {
+                return BadRequest(ex.Message);
```
Comment (jlouros) [inline src/Microservice.Api/Controllers/ImportController.cs]: unexpected error could be server fault, client may not be responsible

File: src/Microservice.Api/Controllers/ProvidersController.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
+using System;
```
Comment (jlouros) [inline src/Microservice.Api/Controllers/ProvidersController.cs]: More than what we asked 👌

File: src/Microservice.Api/Host/BatchFundImportService.cs [hunk 6]
```diff
@@ -84,10 +132,18 @@ private void CheckArchivesToImport()
...
+                        resultsString += r.ErrorMessage +  "\n";
+                }
+
                 // send event to notify indicate completeness
                 HttpClient httpClient = new HttpClient();
-                StringContent sc = new StringContent("process complete for file: " + zipPath);
-                var eventDelivered = httpClient.PostAsync("https://ifl-requestbin.azurewebsites.net/api/sns", sc).Result;
+                StringContent sc = new StringContent("process complete for file: " + zipPath + " " + resultsString);
+                
+                //var eventDelivered = httpClient.PostAsync("https://ifl-requestbin.azurewebsites.net/api/sns", sc).Result;
```
Comment (jlouros) [inline src/Microservice.Api/Host/BatchFundImportService.cs]: why event notification was discarded?

**Comment by jlouros** [comment]
### This candidate was unable to solve the following issues:

- did not use dependency injection
- did not use logging
- unit test setup
- did not create any tests
- strongly type configuration
- multiple file upload will clash (static file name)

**Comment by oranintelliflo** [comment]
- Imports to same file name
- Returns BadRequest when exception is caught. return exception.Message which could reveal internal implementation
- FundsController - Get - If a fund isn't found returns NoContent instead of NotFound
- Hard coded URL for created resoruce
- ProvidersController - Returnds a mixture of ActionResult and data (IEnumerable<Provider>) - attention to details
- CheckArchivesToImport - Code not defensive. Plenty of potential exceptions but no exception handling

**Pull request closed without merging**